deeppavlov/models/preprocessors/ecommerce_preprocess.py

```diff
@@ -0,0 +1,126 @@
+
+@register('ecommerce_preprocess')
+class EcommercePreprocess(Component):
+    """Class to process strings for EcommerceBot skill
+
+    Parameters:
+        spacy_model: SpaCy model name
+        disable: SpaCy pipeline to disable
+    """
+
+    def __init__(self, spacy_model: str = 'en_core_web_sm', disable: Optional[List[str]] = None, **kwargs):
+        if disable is None:
+            disable = ['parser', 'ner']
+
+        self.model = spacy.load(spacy_model, disable=disable)
+
+        below = lambda text: bool(re.compile(r'below|cheap').match(text))
+        BELOW = self.model.vocab.add_flag(below)
+
+        above = lambda text: bool(re.compile(r'above|start').match(text))
+        ABOVE = self.model.vocab.add_flag(above)
+
+        self.matcher = Matcher(self.model.vocab)
+
+        self.matcher.add('below', None, [{BELOW: True}, {'LOWER': 'than', 'OP': '?'},
+                        {'LOWER': 'from', 'OP': '?'}, {'ORTH': '$', 'OP': '?'},
+                        {'ENT_TYPE': 'MONEY', 'LIKE_NUM': True}])
+
+        self.matcher.add('above', None, [{ABOVE: True}, {'LOWER': 'than', 'OP': '?'},
+                        {'LOWER': 'from', 'OP': '?'}, {'ORTH': '$', 'OP': '?'},
+                        {'ENT_TYPE': 'MONEY', 'LIKE_NUM': True}])
+
+    def __call__(self, **kwargs):
+        pass
+
+    def extract_money(self, doc) -> Tuple[List, Tuple[float, float]]:
+        """Extract money entities and money related tokens from `doc`.
+
+        Parameters:
+            doc: a list of tokens with corresponding tags, lemmas, etc.
+
+        Returns:
+            doc_no_money: doc with no money related tokens.
+            money_range: money range from `money_range[0]` to `money_range[1]` extracted from the doc.
+        """
+
+        matches = self.matcher(doc)
+        money_range: Tuple = ()
+        doc_no_money = list(doc)
+        negated = False
+
+        for match_id, start, end in matches:
+            string_id = self.model.vocab.strings[match_id]
+            span = doc[start:end]
+            for child in doc[start].children:
+                if child.dep_ == 'neg':
+                    negated = True
+
+            num_token = [token for token in span if token.like_num == True]
+            if (string_id == 'below' and negated == False) or (string_id == 'above' and negated == True):
+                money_range = (0, float(num_token[0].text))                
+
+            if (string_id == 'above' and negated == False) or (string_id == 'below' and negated == True):
+                money_range = (float(num_token[0].text), float(math.inf))                
+
+            del doc_no_money[start:end+1]
+        return doc_no_money, money_range
+
+    def analyze(self, text: str) -> Iterable:
+        """SpaCy `text` preprocessing"""
+        return self.model(text)
+
+    def spacy2dict(self, doc, fields: List[str] = ['tag_', 'like_num', 'lemma_', 'text']) -> List[Dict[Any, Any]]:
```

Comment:
Doesn't your IDE highlight this?
https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments
doc could be of type spacy.tokens.Doc...